src/Bicep.Core/Workspaces/SourceFileGroupingBuilder.cs

```diff
@@ -161,291 +161,165 @@ private ResultWithDiagnostic<ISourceFile> PopulateRecursive(Uri fileUri, Artifac
         }
 
-        private (ArtifactReference? reference, Result<Uri, UriResolutionError> result) GetArtifactRestoreResult(Uri parentFileUri, IArtifactReferenceSyntax referenceSyntax)
+        private ArtifactResolutionInfo? TryGetImplicitProvider(string providerName, BicepSourceFile file, RootConfiguration config)
         {
-            if (!dispatcher.TryGetArtifactReference(referenceSyntax, parentFileUri).IsSuccess(out var artifactReference, out var referenceResolutionError))
+            if (!config.ProvidersConfig.TryGetProviderSource(providerName).IsSuccess(out var providerEntry, out var errorBuilder))
+            {
+                return new(file, null, null, new(errorBuilder), RequiresRestore: false);
+            }
+
+            if (providerEntry.BuiltIn)
+            {
+                return null;
+            }
+
+            if (!OciArtifactReference.TryParse(ArtifactType.Provider, null, $"{providerEntry.Source}:{providerEntry.Version}", config, file.FileUri).IsSuccess(out var artifactReference, out errorBuilder))
+            {
+                // reference is not valid
+                return new(file, null, null, new(errorBuilder), RequiresRestore: false);
+            }
+
+            var (result, requiresRestore) = GetArtifactRestoreResult(artifactReference);
+            return new(file, null, artifactReference, result, RequiresRestore: requiresRestore);
+        }
+
+        private ArtifactResolutionInfo GetArtifactRestoreResult(BicepSourceFile sourceFile, IArtifactReferenceSyntax referenceSyntax)
+        {
+            if (!dispatcher.TryGetArtifactReference(referenceSyntax, sourceFile.FileUri).IsSuccess(out var artifactReference, out var errorBuilder))
             {
                 // module reference is not valid
-                return (null, new(new UriResolutionError(referenceResolutionError, RequiresRestore: false)));
+                return new(sourceFile, referenceSyntax, null, new(errorBuilder), RequiresRestore: false);
             }
-            return (artifactReference, GetArtifactRestoreResult(artifactReference));
+
+            var (result, requiresRestore) = GetArtifactRestoreResult(artifactReference);
+
+            return new(sourceFile, referenceSyntax, artifactReference, result, RequiresRestore: requiresRestore);
         }
 
-        private Result<Uri, UriResolutionError> GetArtifactRestoreResult(ArtifactReference artifactReference)
+        private (ResultWithDiagnostic<Uri> result, bool requiresRestore) GetArtifactRestoreResult(ArtifactReference artifactReference)
         {
             if (!dispatcher.TryGetLocalArtifactEntryPointUri(artifactReference).IsSuccess(out var artifactFileUri, out var artifactGetPathFailureBuilder))
             {
-                return new(new UriResolutionError(artifactGetPathFailureBuilder, RequiresRestore: false));
+                return (new(artifactGetPathFailureBuilder), requiresRestore: false);
             }
 
             if (forceRestore)
```

Comment:
Should we evaluate this if-clause in the beginning of this function? If `forceRestore` is set, we don't need to try to get the URI (LOC 256-259)

Reply:
No, because this will result in a degraded user experience (they'll be shown a generic "artifact couldn't be restored" message, rather than a message to indicate that there's a problem)